src/template/js/sdk/Capabilities/index.mjs

```diff
@@ -18,43 +18,82 @@
 
 import Transport from '../Transport/index.mjs'
 import Events from '../Events/index.mjs'
-const validProviderMethods = {}
 
-export const registerProviderMethods = (module, methods) => {
-  validProviderMethods[module.toLowerCase()] = methods.concat()
-}
+const providerInterfaces = {}
+
+export const registerProviderInterface = (capability, module, methods) => {
+  if (providerInterfaces[capability]) {
+    throw `Capability ${capability} has multiple provider interfaces registered.`
+  }
 
-const getProviderArgs = function(...args) {
-  const provider = args.pop()
-  const module = args[0].toLowerCase() || '*'
-  return [module, provider]
+  methods.forEach(m => m.name = `${module}.${m.name}`)
+  providerInterfaces[capability] = methods.concat()
 }
 
-const provide = function(...args) {
-  const [module, provider] = getProviderArgs(...args)
-  const methods = Object.getOwnPropertyNames(Object.getPrototypeOf(provider)).filter(item => {
-    return typeof provider[item] === 'function' && item != 'constructor'
-  })
-  const pms = validProviderMethods[module]
-  for (let i = 0; i < methods.length; i++) {
-    const name = methods[i].charAt(0).toUpperCase() + methods[i].slice(1);
-    if (pms.indexOf(methods[i]) !== -1) {
-      Events.listen(module, 'request' + name, async function (req) {
-        const fn = provider[methods[i]]
-        const providerCallArgs = [req.request, resp => {
-          Transport.send(module, methods[i] + 'Response', {
-            correlationId: req.correlationId,
-            response: resp
+const provide = function(capability, provider) {
+  const methods = []
+  const iface = providerInterfaces[capability]
+
+  if (provider.constructor.name !== 'Object') {
+    methods.push(...Object.getOwnPropertyNames(Object.getPrototypeOf(provider)).filter(item => typeof provider[item] === 'function' && item !== 'constructor'))
+  }
+  else {
+    methods.push(...Object.getOwnPropertyNames(provider).filter(item => typeof provider[item] === 'function'))
+  }
+
+  if (iface) {
+
+    // make sure every interfaced method exists in the providers methods list
+    const valid = iface.every(method => methods.find(m => m === method.name.split('.').pop()))
+
+    if (valid) {
+      iface.forEach(imethod => {
+        const parts = imethod.name.split('.')
+        const method = parts.pop();
+        const module = parts.pop().toLowerCase();
+
+        const defined = !!methods.find(m => m === method) 
+
+        if (defined) {
+          Events.listen(module, `request${method.charAt(0).toUpperCase() + method.substr(1)}`, function (request) {
+
+            const providerCallArgs = []
+            
+            // only pass in parameters object if schema exists
+            if (imethod.parameters) {
```

Comment:
I don't think we want to change the number of args the provider method takes.
Seems like it should always take first arg as parameters. If no parameters are necessary this can just be null or empty object.

Reply:
yeah, I had it that way, and it felt weird to pass in `null`.

We're not really "changing" it. This code is just reusable code that all providers leverage. Once we have a release version of the schemas, we can't change it w/out a major version bump, so it felt cleaner to not add the noise of a `null` parameter that would never be useful.

Reply:
I meant it is odd that the author of the provider has to have knowledge of the API to create the method signature.
The signature could be:
```
my_method(parameters, handshake)
```
OR
```
my_method(handshake)
```
OR
```
my_method(parameters)
```
OR
```
my_method()
```
Depending on it's openrpc definition.
I feel a consistent API is better:
```
my_method(parameters, handshake)
```
Where parameters can just be null or empty object if it has no params.

Reply:
Yeah, you've convinced me that parameters should always be there. So we'll have one signature:

```typescript
function my_method(parameters: object, handshake?: function)
```

parameters can be null if there are none, and handshake can be left off when it doesn't make sense.